src/System.Net.Requests/src/System/Net/WebClient.cs

```diff
@@ -0,0 +1,2127 @@
+                }
+            }
+
+            return true;
+        }
+
+        private static readonly char[] s_parseContentTypeSeparators = new char[] { ';', '=', ' ' };
+        private static readonly Encoding[] s_knownEncodings = { Encoding.UTF8, Encoding.UTF32, Encoding.Unicode, Encoding.BigEndianUnicode };
+
+        private string GetStringUsingEncoding(WebRequest request, byte[] data)
+        {
+            Encoding enc = null;
+            int bomLengthInData = -1;
+
+            // Figure out encoding by first checking for encoding string in Content-Type HTTP header
+            // This can throw NotImplementedException if the derived class of WebRequest doesn't support it.
+            string contentType;
+            try
+            {
+                contentType = request.ContentType;
+            }
+            catch (Exception e) when (e is NotImplementedException || e is NotSupportedException) // some types do this
+            {
+                contentType = null;
+            }
+
+            if (contentType != null)
+            {
+                contentType = contentType.ToLower(CultureInfo.InvariantCulture);
+                string[] parsedList = contentType.Split(s_parseContentTypeSeparators);
+                bool nextItem = false;
+                foreach (string item in parsedList)
+                {
+                    if (item == "charset")
+                    {
+                        nextItem = true;
+                    }
+                    else if (nextItem)
+                    {
+                        try
+                        {
+                            enc = Encoding.GetEncoding(item);
+                        }
+                        catch (ArgumentException)
+                        {
+                            // Eat ArgumentException here.    
+                            // We'll assume that Content-Type encoding might have been garbled and wasn't present at all.
+                            break;
+                        }
+                        // Unexpected exceptions are thrown back to caller
+                    }
+                }
+            }
+
+            // If no content encoding listed in the ContentType HTTP header, or no Content-Type header present, then
+            // check for a byte-order-mark (BOM) in the data to figure out encoding.
+            if (enc == null)
+            {
+                // UTF32 must be tested before Unicode because it's BOM is the same but longer.
+                Encoding[] encodings = s_knownEncodings;
+                for (int i = 0; i < encodings.Length; i++)
+                {
+                    byte[] preamble = encodings[i].GetPreamble();
+                    if (ByteArrayHasPrefix(preamble, data))
+                    {
+                        enc = encodings[i];
+                        bomLengthInData = preamble.Length;
+                        break;
+                    }
+                }
+            }
```

Comment:
Yes.  If that were refactored to be reusable, we could use it here subsequently. I've explicitly not rewritten large segments (other than the async methods I mentioned) nor tried to optimize such things in this PR.